pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -28,37 +29,53 @@ import (
 // Node is a set of constraints, compatible pods, and possible instance types that could fulfill these constraints. This
 // will be turned into one or more actual node instances within the cluster after bin packing.
 type Node struct {
+	Name                string
 	Constraints         *v1alpha5.Constraints
 	InstanceTypeOptions []cloudprovider.InstanceType
 	Pods                []*v1.Pod
 
 	requests v1.ResourceList
 }
 
+var nodeID int64
+
 func NewNode(constraints *v1alpha5.Constraints, daemonResources v1.ResourceList, instanceTypes []cloudprovider.InstanceType) *Node {
-	return &Node{
+	n := &Node{
+		Name:                fmt.Sprintf("hostname-placeholder-%04d", atomic.AddInt64(&nodeID, 1)),
 		Constraints:         constraints.DeepCopy(),
 		InstanceTypeOptions: instanceTypes,
 		requests:            daemonResources,
 	}
-}
 
-func (n *Node) Add(pod *v1.Pod) error {
-	podRequirements := v1alpha5.NewPodRequirements(pod)
+	n.Constraints.Requirements = n.Constraints.Requirements.Add(v1.NodeSelectorRequirement{
+		Key:      v1.LabelHostname,
+		Operator: v1.NodeSelectorOpIn,
+		Values:   []string{n.Name},
+	})
+	return n
+}
 
-	if len(n.Pods) != 0 {
-		// TODO: remove this check for n.Pods once we properly support hostname topology spread
-		if err := n.Constraints.Requirements.Compatible(podRequirements); err != nil {
-			return err
-		}
+func (n *Node) Add(topology *Topology, p *v1.Pod) error {
```

Comment:
Seems fine, probably held off because constructor has a bunch of params already, but agree that Add() is looking awkward with this topology parameter.